File: src/oca/java/src/org/opennebula/client/vnet/VirtualNetwork.java
/*******************************************************************************
 * Copyright 2002-2015, OpenNebula Project (OpenNebula.org), C12G Labs
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ******************************************************************************/
package org.opennebula.client.vnet;


import org.opennebula.client.Client;
import org.opennebula.client.OneResponse;
import org.opennebula.client.PoolElement;
import org.w3c.dom.Node;

/**
 * This class represents an OpenNebula virtual network.
 * It also offers static XML-RPC call wrappers.
 */
public class VirtualNetwork extends PoolElement{

    private static final String METHOD_PREFIX   = "vn.";
    private static final String ALLOCATE        = METHOD_PREFIX + "allocate";
    private static final String INFO            = METHOD_PREFIX + "info";
    private static final String DELETE          = METHOD_PREFIX + "delete";
    private static final String ADDAR           = METHOD_PREFIX + "add_ar";
    private static final String RMAR            = METHOD_PREFIX + "rm_ar";
    private static final String UPDATEAR        = METHOD_PREFIX + "update_ar";
    private static final String CHOWN           = METHOD_PREFIX + "chown";
    private static final String CHMOD           = METHOD_PREFIX + "chmod";
    private static final String UPDATE          = METHOD_PREFIX + "update";
    private static final String HOLD            = METHOD_PREFIX + "hold";
    private static final String RELEASE         = METHOD_PREFIX + "release";
    private static final String RENAME          = METHOD_PREFIX + "rename";
    private static final String RESERVE         = METHOD_PREFIX + "reserve";
    private static final String FREEAR          = METHOD_PREFIX + "free_ar";

    /**
     * Creates a new virtual network representation.
     *
     * @param id The virtual network id (nid) .
     * @param client XML-RPC Client.
     */
    public VirtualNetwork(int id, Client client)
    {
        super(id, client);
    }

    /**
     * @see PoolElement
     */
    protected VirtualNetwork(Node xmlElement, Client client)
    {
        super(xmlElement, client);
    }

    // =================================
    // Static XML-RPC methods
    // =================================

    /**
     * Allocates a new virtual network in OpenNebula.
     *
     * @param client XML-RPC Client.
     * @param description A string containing the template
     * of the virtual network.
     * @param clusterId The cluster ID. If it is -1, this virtual network
     * won't be added to any cluster.
     *
     * @return If successful the message contains the associated
     * id generated for this virtual network.
     */
    public static OneResponse allocate(
            Client  client,
            String  description,
            int     clusterId)
    {
        return client.call(ALLOCATE, description, clusterId);
    }

    /**
     * Allocates a new virtual network in OpenNebula.
     *
     * @param client XML-RPC Client.
     * @param description A string containing the template
     * of the virtual network.
     *
     * @return If successful the message contains the associated
     * id generated for this virtual network.
     */
    public static OneResponse allocate(
            Client  client,
            String  description)
    {
        return allocate(client, description, -1);
    }

    /**
     * Retrieves the information of the given virtual network
     *
     * @param client XML-RPC Client.
     * @param id the virtual network id (nid) for the network to
     * retrieve the information from.
     * @return If successful the message contains the string
     * with the information returned by OpenNebula.
     */
    public static OneResponse info(Client client, int id)
    {
        return client.call(INFO, id);
    }

    /**
     * Deletes a network from OpenNebula.
     *
     * @param client XML-RPC Client.
     * @param id The virtual network id (nid) of the target network.
     * @return A encapsulated response.
     */
    public static OneResponse delete(Client client, int id)
    {
        return client.call(DELETE, id);
    }

    /**
     * Publishes or unpublishes a virtual network.
     *
     * @param client XML-RPC Client.
     * @param id The virtual network id (nid) of the target network.
     * @param publish True for publishing, false for unpublishing.
     * @return If successful the message contains the image id.
     */
    public static OneResponse publish(Client client, int id, boolean publish)
    {
        int group_u = publish ? 1 : 0;

        return chmod(client, id, -1, -1, -1, group_u, -1, -1, -1, -1, -1);
    }

    /**
     * Adds an Address Range to the VirtualNetwork
     *
     * @param client XML-RPC Client.
     * @param id The virtual network id (nid) of the target network.
     * @param template AR to add, example:
     *      AR = [
     *          TYPE = IP4,
     *          IP = 192.168.0.5,
     *          SIZE = 10 ]
     *
     * @return A encapsulated response.
     */
    public static OneResponse addAr(Client client, int id, String template)
    {
        return client.call(ADDAR, id, template);
    }

    /**
     * Removes an Address Range from the VirtualNetwork
     *
     * @param client XML-RPC Client.
     * @param id The virtual network id (nid) of the target network.
     * @param arId Id of the Address Range to remove
     * @return A encapsulated response.
     */
    public static OneResponse rmAr(Client client, int id, int arId)
    {
        return client.call(RMAR, id, arId);
    }

    /**
     * Upates an Address Range from the VirtualNetwork
     *
     * @param client XML-RPC Client.
     * @param id The virtual network id (nid) of the target network.
     * @param template AR to update, example:
     *      AR = [
     *          AR_ID = 3,
     *          TYPE  = IP4,
     *          IP    = 192.168.0.5,
     *          SIZE  = 10 ]
     *
     * @return A encapsulated response.
     */
    public static OneResponse updateAr(Client client, int id, String template)
    {
        return client.call(UPDATEAR, id, template);
    }

    /**
     * Holds a VirtualNetwork lease, marking it as used
     *
     * @param client XML-RPC Client.
     * @param id The virtual network id (nid) of the target network.
     * @param template Address to hold, examples:
     *<pre>
     *      LEASES = [ IP = 192.168.0.5 ]
     *      LEASES = [ MAC = 02:00:0a:00:00:96 ]
     *      LEASES = [ IP = 192.168.0.5, AR_ID = 3 ]
     *</pre>
     * @return A encapsulated response.
     */
    public static OneResponse hold(Client client, int id, String template)
    {
        return client.call(HOLD, id, template);
    }

    /**
     * Releases a VirtualNetwork lease on hold
     *
     * @param client XML-RPC Client.
     * @param id The virtual network id (nid) of the target network.
     * @param template Address to release, examples:
     *<pre>
     *      LEASES = [ IP = 192.168.0.5 ]
     *      LEASES = [ MAC = 02:00:0a:00:00:96 ]
     *      LEASES = [ IP = 192.168.0.5, AR_ID = 3 ]
     *</pre>
     * @return A encapsulated response.
     */
    public static OneResponse release(Client client, int id, String template)
    {
        return client.call(RELEASE, id, template);
    }

    /**
     * Changes the owner/group
     *
     * @param client XML-RPC Client.
     * @param id The virtual network id (nid) of the target network.
     * @param uid The new owner user ID. Set it to -1 to leave the current one.
     * @param gid The new group ID. Set it to -1 to leave the current one.
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse chown(Client client, int id, int uid, int gid)
    {
        return client.call(CHOWN, id, uid, gid);
    }

    /**
     * Changes the VirtualNetwork permissions
     *
     * @param client XML-RPC Client.
     * @param id The virtual network id (nid) of the target network.
     * @param owner_u 1 to allow, 0 deny, -1 do not change
     * @param owner_m 1 to allow, 0 deny, -1 do not change
     * @param owner_a 1 to allow, 0 deny, -1 do not change
     * @param group_u 1 to allow, 0 deny, -1 do not change
     * @param group_m 1 to allow, 0 deny, -1 do not change
     * @param group_a 1 to allow, 0 deny, -1 do not change
     * @param other_u 1 to allow, 0 deny, -1 do not change
     * @param other_m 1 to allow, 0 deny, -1 do not change
     * @param other_a 1 to allow, 0 deny, -1 do not change
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse chmod(Client client, int id,
                                    int owner_u, int owner_m, int owner_a,
                                    int group_u, int group_m, int group_a,
                                    int other_u, int other_m, int other_a)
    {
        return chmod(client, CHMOD, id,
                owner_u, owner_m, owner_a,
                group_u, group_m, group_a,
                other_u, other_m, other_a);
    }

    /**
     * Changes the permissions
     *
     * @param client XML-RPC Client.
     * @param id The id of the target object.
     * @param octet Permissions octed , e.g. 640
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse chmod(Client client, int id, String octet)
    {
        return chmod(client, CHMOD, id, octet);
    }

    /**
     * Changes the permissions
     *
     * @param client XML-RPC Client.
     * @param id The id of the target object.
     * @param octet Permissions octed , e.g. 640
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse chmod(Client client, int id, int octet)
    {
        return chmod(client, CHMOD, id, octet);
    }

    /**
     * Replaces the VirtualNetwork template contents.
     *
     * @param client XML-RPC Client.
     * @param id The vnet id of the target vnet we want to modify.
     * @param new_template New template contents.
     * @param append True to append new attributes instead of replace the whole template
     * @return If successful the message contains the vnet id.
     */
    public static OneResponse update(Client client, int id, String new_template,
        boolean append)
    {
        return client.call(UPDATE, id, new_template, append ? 1 : 0);
    }

    /**
     * Renames this VirtualNetwork
     *
     * @param client XML-RPC Client.
     * @param id The VirtualNetwork id of the target VirtualNetwork.
     * @param name New name for the VirtualNetwork.
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse rename(Client client, int id, String name)
    {
        return client.call(RENAME, id, name);
    }

    /**
     * Reserve a set of addresses from this virtual network
     *
     * @param client XML-RPC Client.
     * @param id The virtual network id (nid) of the target network.
     * @param template of the reservation. Examples:
     *<pre>
     *      SIZE  = 10
     *
     *      SIZE  = 10
     *      AR_ID = 3
     *      NAME  = "new_network"
     *
     *      SIZE  = 10
     *      IP    = 192.168.10.50
     *      NETWORK_ID = 9
     *</pre>
     * @return A encapsulated response.
     */
    public static OneResponse reserve(Client client, int id, String template)
    {
        return client.call(RESERVE, id, template);
    }

    /**
     * Removes an Address Range from the VirtualNetwork
     *
     * @param client XML-RPC Client.
     * @param id The virtual network id (nid) of the target network.
     * @param arId Id of the Address Range to remove
     * @return A encapsulated response.
     */
    public static OneResponse free(Client client, int id, int arId)
    {
        return client.call(FREEAR, id, arId);
    }

    // =================================
    // Instanced object XML-RPC methods
    // =================================

    /**
     * Loads the xml representation of the virtual network.
     * The info is also stored internally.
     *
     * @see VirtualNetwork#info(Client, int)
     */
    public OneResponse info()
    {
        OneResponse response = info(client, id);
        super.processInfo(response);
        return response;
    }

    /**
     * Deletes the network from OpenNebula.
     *
     * @return A encapsulated response.
     */
    public OneResponse delete()
    {
        return delete(client, id);
    }

    /**
     * Publishes or unpublishes the virtual network.
     *
     * @param publish True for publishing, false for unpublishing.
     * @return If successful the message contains the image id.
     */
    public OneResponse publish(boolean publish)
    {
        return publish(client, id, publish);
    }

    /**
     * Publishes the virtual network.
     *
     * @return If successful the message contains the image id.
     */
    public OneResponse publish()
    {
        return publish(true);
    }

    /**
     * Unpublishes the virtual network.
     *
     * @return If successful the message contains the image id.
     */
    public OneResponse unpublish()
    {
        return publish(false);
    }

    /**
     * Adds an Address Range to the VirtualNetwork
     *
     * @param template AR to add, example:
     *<pre>
     *      AR = [
     *          TYPE = IP4,
     *          IP = 192.168.0.5,
     *          SIZE = 10 ]
     *</pre>
     *
     * @return A encapsulated response.
     */
    public OneResponse addAr(String template)
    {
        return addAr(client, id, template);
    }

    /**
     * Removes an Address Range from the VirtualNetwork
     *
     * @param arId Id of the Address Range to remove
     * @return A encapsulated response.
     */
    public OneResponse rmAr(int arId)
    {
        return rmAr(client, id, arId);
    }

    /**
     * Upates an Address Range from the VirtualNetwork
     *
     * @param template AR to update, example:
     *<pre>
     *      AR = [
     *          AR_ID = 3,
     *          TYPE  = IP4,
     *          IP    = 192.168.0.5,
     *          SIZE  = 10 ]
     *</pre>
     *
     * @return A encapsulated response.
     */
    public OneResponse updateAr(String template)
    {
        return updateAr(client, id, template);
    }

    /**
     * Holds a VirtualNetwork lease, marking it as used
     *
     * @param ip IP or MAC to hold, e.g. "192.168.0.5", "02:00:0a:00:00:96"
     * @return A encapsulated response.
     */
    public OneResponse hold(String ip)
    {
        String addr_name = ip.contains(":") ? "MAC" : "IP";

        String lease_template = "LEASES = [ "+addr_name+" = "+ip+"]";

        return hold(client, id, lease_template);
    }

    /**
     * Holds a VirtualNetwork lease, marking it as used
     *
     * @param ip IP or MAC to hold, e.g. "192.168.0.5", "02:00:0a:00:00:96"
     * @param arId Id of the Address Range to hold the lease from
     * @return A encapsulated response.
     */
    public OneResponse hold(String ip, int arId)
    {
        String addr_name = ip.contains(":") ? "MAC" : "IP";

        String lease_template =
            "LEASES = [ "+addr_name+" = "+ip+", AR_ID = "+arId+" ]";

        return hold(client, id, lease_template);
    }

    /**
     * Releases a VirtualNetwork lease on hold
     *
     * @param ip IP or MAC to hold, e.g. "192.168.0.5", "02:00:0a:00:00:96"
     * @return A encapsulated response.
     */
    public OneResponse release(String ip)
    {
        String addr_name = ip.contains(":") ? "MAC" : "IP";

        String lease_template = "LEASES = [ "+addr_name+" = "+ip+"]";

        return release(client, id, lease_template);
    }

    /**
     * Releases a VirtualNetwork lease on hold
     *
     * @param ip IP or MAC to hold, e.g. "192.168.0.5", "02:00:0a:00:00:96"
     * @param arId Id of the Address Range to release the lease from
     * @return A encapsulated response.
     */
    public OneResponse release(String ip, int arId)
    {
        String addr_name = ip.contains(":") ? "MAC" : "IP";

        String lease_template =
            "LEASES = [ "+addr_name+" = "+ip+", AR_ID = "+arId+" ]";

        return release(client, id, lease_template);
    }

    /**
     * Changes the owner/group
     *
     * @param uid The new owner user ID. Set it to -1 to leave the current one.
     * @param gid The new group ID. Set it to -1 to leave the current one.
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse chown(int uid, int gid)
    {
        return chown(client, id, uid, gid);
    }

    /**
     * Changes the owner
     *
     * @param uid The new owner user ID.
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse chown(int uid)
    {
        return chown(uid, -1);
    }

    /**
     * Changes the group
     *
     * @param gid The new group ID.
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse chgrp(int gid)
    {
        return chown(-1, gid);
    }

    /**
     * Changes the VirtualNetwork permissions
     *
     * @param owner_u 1 to allow, 0 deny, -1 do not change
     * @param owner_m 1 to allow, 0 deny, -1 do not change
     * @param owner_a 1 to allow, 0 deny, -1 do not change
     * @param group_u 1 to allow, 0 deny, -1 do not change
     * @param group_m 1 to allow, 0 deny, -1 do not change
     * @param group_a 1 to allow, 0 deny, -1 do not change
     * @param other_u 1 to allow, 0 deny, -1 do not change
     * @param other_m 1 to allow, 0 deny, -1 do not change
     * @param other_a 1 to allow, 0 deny, -1 do not change
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse chmod(int owner_u, int owner_m, int owner_a,
                             int group_u, int group_m, int group_a,
                             int other_u, int other_m, int other_a)
    {
        return chmod(client, id,
                    owner_u, owner_m, owner_a,
                    group_u, group_m, group_a,
                    other_u, other_m, other_a);
    }

    /**
     * Changes the permissions
     *
     * @param octet Permissions octed , e.g. 640
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse chmod(String octet)
    {
        return chmod(client, id, octet);
    }

    /**
     * Changes the permissions
     *
     * @param octet Permissions octed , e.g. 640
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse chmod(int octet)
    {
        return chmod(client, id, octet);
    }

    /**
     * Replaces the VirtualNetwork template contents.
     *
     * @param new_template New template contents.
     * @return If successful the message contains the vnet id.
     */
    public OneResponse update(String new_template)
    {
        return update(new_template, false);
    }

    /**
     * Replaces the VirtualNetwork template contents.
     *
     * @param new_template New template contents.
     * @param append True to append new attributes instead of replace the whole template
     * @return If successful the message contains the vnet id.
     */
    public OneResponse update(String new_template, boolean append)
    {
        return update(client, id, new_template, append);
    }

    /**
     * Renames this VirtualNetwork
     *
     * @param name New name for the VirtualNetwork.
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse rename(String name)
    {
        return rename(client, id, name);
    }

    /**
     * Reserve a set of addresses from this virtual network
     *
     * @param template of the reservation. Examples:
     *<pre>
     *      SIZE  = 10
     *
     *      SIZE  = 10
     *      AR_ID = 3
     *      NAME  = "new_network"
     *
     *      SIZE  = 10
     *      IP    = 192.168.10.50
     *      NETWORK_ID = 9
     *</pre>
     * @return A encapsulated response.
     */
    public OneResponse reserve(String template)
    {
        return reserve(client, id, template);
    }

    /**
     * Removes an Address Range from the VirtualNetwork
     *
     * @param arId Id of the Address Range to remove
     * @return A encapsulated response.
     */
    public OneResponse free(int arId)
    {
        return free(client, id, arId);
    }

    // =================================
    // Helpers
    // =================================

}

File: src/oca/go/src/goca/virtualnetwork.go
/* -------------------------------------------------------------------------- */
/* Copyright 2002-2024, OpenNebula Project, OpenNebula Systems                */
/*                                                                            */
/* Licensed under the Apache License, Version 2.0 (the "License"); you may    */
/* not use this file except in compliance with the License. You may obtain    */
/* a copy of the License at                                                   */
/*                                                                            */
/* http://www.apache.org/licenses/LICENSE-2.0                                 */
/*                                                                            */
/* Unless required by applicable law or agreed to in writing, software        */
/* distributed under the License is distributed on an "AS IS" BASIS,          */
/* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   */
/* See the License for the specific language governing permissions and        */
/* limitations under the License.                                             */
/*--------------------------------------------------------------------------- */

package goca

import (
	"context"
	"encoding/xml"
	"errors"

	"github.com/OpenNebula/one/src/oca/go/src/goca/parameters"
	"github.com/OpenNebula/one/src/oca/go/src/goca/schemas/shared"
	vn "github.com/OpenNebula/one/src/oca/go/src/goca/schemas/virtualnetwork"
)

// VirtualNetworksController is a controller for a pool of VirtualNetworks
type VirtualNetworksController entitiesController

// VirtualNetworkController is a controller for VirtualNetwork entities
type VirtualNetworkController entityController

// VirtualNetworks returns a VirtualNetworks controller
func (c *Controller) VirtualNetworks() *VirtualNetworksController {
	return &VirtualNetworksController{c}
}

// VirtualNetwork returns a VirtualNetwork controller
func (c *Controller) VirtualNetwork(id int) *VirtualNetworkController {
	return &VirtualNetworkController{c, id}
}

// ByName returns a VirtualNetwork ID from name
func (c *VirtualNetworksController) ByName(name string, args ...int) (int, error) {
	return c.ByNameContext(context.Background(), name, args...)
}

// ByNameContext returns a VirtualNetwork ID from name
func (c *VirtualNetworksController) ByNameContext(ctx context.Context, name string, args ...int) (int, error) {
	var id int

	virtualNetworkPool, err := c.InfoContext(ctx, args...)
	if err != nil {
		return -1, err
	}

	match := false
	for i := 0; i < len(virtualNetworkPool.VirtualNetworks); i++ {
		if virtualNetworkPool.VirtualNetworks[i].Name != name {
			continue
		}
		if match {
			return -1, errors.New("multiple resources with that name")
		}
		id = virtualNetworkPool.VirtualNetworks[i].ID
		match = true
	}
	if !match {
		return -1, errors.New("resource not found")
	}

	return id, nil
}

// Info returns a virtualnetwork pool.
func (vc *VirtualNetworksController) Info(args ...int) (*vn.Pool, error) {
	return vc.InfoContext(context.Background(), args...)
}

// InfoContext returns a virtualnetwork pool.
func (vc *VirtualNetworksController) InfoContext(ctx context.Context, args ...int) (*vn.Pool, error) {
	var who, start, end int

	switch len(args) {
	case 0:
		who = parameters.PoolWhoAll
		start = -1
		end = -1
	case 1:
		who = args[0]
		start = -1
		end = -1
	case 3:
		who = args[0]
		start = args[1]
		end = args[2]
	default:
		return nil, errors.New("Wrong number of arguments")
	}

	response, err := vc.c.Client.CallContext(ctx, "one.vnpool.info", who, start, end)
	if err != nil {
		return nil, err
	}

	vnPool := &vn.Pool{}
	err = xml.Unmarshal([]byte(response.Body()), &vnPool)
	if err != nil {
		return nil, err
	}

	return vnPool, nil
}

// Info retrieves information for the virtual network.
func (vc *VirtualNetworkController) Info(decrypt bool) (*vn.VirtualNetwork, error) {
	return vc.InfoContext(context.Background(), decrypt)
}

// InfoContext retrieves information for the virtual network.
func (vc *VirtualNetworkController) InfoContext(ctx context.Context, decrypt bool) (*vn.VirtualNetwork, error) {
	response, err := vc.c.Client.CallContext(ctx, "one.vn.info", vc.ID, decrypt)
	if err != nil {
		return nil, err
	}
	vn := &vn.VirtualNetwork{}
	err = xml.Unmarshal([]byte(response.Body()), vn)
	if err != nil {
		return nil, err
	}
	return vn, nil
}

// Create allocates a new virtualnetwork. It returns the new virtualnetwork ID.
// * tpl: template of the virtualnetwork
// * clusterID: The cluster ID. If it is -1, the default one will be used.
func (vc *VirtualNetworksController) Create(tpl string, clusterID int) (int, error) {
	return vc.CreateContext(context.Background(), tpl, clusterID)
}

// CreateContext allocates a new virtualnetwork. It returns the new virtualnetwork ID.
// * ctx: context for cancelation
// * tpl: template of the virtualnetwork
// * clusterID: The cluster ID. If it is -1, the default one will be used.
func (vc *VirtualNetworksController) CreateContext(ctx context.Context, tpl string, clusterID int) (int, error) {
	response, err := vc.c.Client.CallContext(ctx, "one.vn.allocate", tpl, clusterID)
	if err != nil {
		return -1, err
	}

	return response.BodyInt(), nil
}

// Delete deletes the given virtual network from the pool.
func (vc *VirtualNetworkController) Delete() error {
	return vc.DeleteContext(context.Background())
}

// DeleteContext deletes the given virtual network from the pool.
func (vc *VirtualNetworkController) DeleteContext(ctx context.Context) error {
	_, err := vc.c.Client.CallContext(ctx, "one.vn.delete", vc.ID)
	return err
}

// AddAR adds address ranges to a virtual network.
// * tpl: template of the address ranges to add. Syntax can be the usual attribute=value or XML
func (vc *VirtualNetworkController) AddAR(tpl string) error {
	return vc.AddARContext(context.Background(), tpl)
}

// AddARContext adds address ranges to a virtual network.
// * ctx: context for cancelation
// * tpl: template of the address ranges to add. Syntax can be the usual attribute=value or XML
func (vc *VirtualNetworkController) AddARContext(ctx context.Context, tpl string) error {
	_, err := vc.c.Client.CallContext(ctx, "one.vn.add_ar", vc.ID, tpl)
	return err
}

// RmAR removes an address range from a virtual network.
// * arID: ID of the address range to remove.
func (vc *VirtualNetworkController) RmAR(arID int) error {
	return vc.RmARContext(context.Background(), arID)
}

// RmARContext removes an address range from a virtual network.
// * ctx: context for cancelation
// * arID: ID of the address range to remove.
func (vc *VirtualNetworkController) RmARContext(ctx context.Context, arID int) error {
	_, err := vc.c.Client.CallContext(ctx, "one.vn.rm_ar", vc.ID, arID)
	return err
}

// UpdateAR updates the attributes of an address range.
// * tpl: template of the address ranges to update. Syntax can be the usual attribute=value or XML
func (vc *VirtualNetworkController) UpdateAR(tpl string) error {
	return vc.UpdateARContext(context.Background(), tpl)
}

// UpdateARContext updates the attributes of an address range.
// * ctx: context for cancelation
// * tpl: template of the address ranges to update. Syntax can be the usual attribute=value or XML
func (vc *VirtualNetworkController) UpdateARContext(ctx context.Context, tpl string) error {
	_, err := vc.c.Client.CallContext(ctx, "one.vn.update_ar", vc.ID, tpl)
	return err
}

// Reserve reserve network addresses. It returns the Reserved Virtual Network ID
// * tpl: Template
func (vc *VirtualNetworkController) Reserve(tpl string) (int, error) {
	return vc.ReserveContext(context.Background(), tpl)
}

// ReserveContext reserve network addresses. It returns the Reserved Virtual Network ID
// * ctx: context for cancelation
// * tpl: Template
func (vc *VirtualNetworkController) ReserveContext(ctx context.Context, tpl string) (int, error) {
	response, err := vc.c.Client.CallContext(ctx, "one.vn.reserve", vc.ID, tpl)
	if err != nil {
		return -1, err
	}

	return response.BodyInt(), nil
}

// FreeAR frees a reserved address range from a virtual network.
// * arID: ID of the address range to free.
func (vc *VirtualNetworkController) FreeAR(arID int) error {
	return vc.FreeARContext(context.Background(), arID)
}

// FreeARContext frees a reserved address range from a virtual network.
// * ctx: context for cancelation
// * arID: ID of the address range to free.
func (vc *VirtualNetworkController) FreeARContext(ctx context.Context, arID int) error {
	_, err := vc.c.Client.CallContext(ctx, "one.vn.free_ar", vc.ID, arID)
	return err
}

// Hold holds a virtual network Lease as used.
// * tpl: template of the lease to hold
func (vc *VirtualNetworkController) Hold(tpl string) error {
	return vc.HoldContext(context.Background(), tpl)
}

// HoldContext holds a virtual network Lease as used.
// * ctx: context for cancelation
// * tpl: template of the lease to hold
func (vc *VirtualNetworkController) HoldContext(ctx context.Context, tpl string) error {
	_, err := vc.c.Client.CallContext(ctx, "one.vn.hold", vc.ID, tpl)
	return err
}

// Release releases a virtual network Lease on hold.
// * tpl: template of the lease to release
func (vc *VirtualNetworkController) Release(tpl string) error {
	return vc.ReleaseContext(context.Background(), tpl)
}

// ReleaseContext releases a virtual network Lease on hold.
// * ctx: context for cancelation
// * tpl: template of the lease to release
func (vc *VirtualNetworkController) ReleaseContext(ctx context.Context, tpl string) error {
	_, err := vc.c.Client.CallContext(ctx, "one.vn.release", vc.ID, tpl)
	return err
}

// Update adds virtual network content.
// * tpl: The new virtual network contents. Syntax can be the usual attribute=value or XML.
// * uType: Update type: Replace: Replace the whole template.
//   Merge: Merge new template with the existing one.
func (vc *VirtualNetworkController) Update(tpl string, uType parameters.UpdateType) error {
	return vc.UpdateContext(context.Background(), tpl, uType)
}

// UpdateContext adds virtual network content.
//   - ctx: context for cancelation
//   - tpl: The new virtual network contents. Syntax can be the usual attribute=value or XML.
//   - uType: Update type: Replace: Replace the whole template.
//     Merge: Merge new template with the existing one.
func (vc *VirtualNetworkController) UpdateContext(ctx context.Context, tpl string, uType parameters.UpdateType) error {
	_, err := vc.c.Client.CallContext(ctx, "one.vn.update", vc.ID, tpl, uType)
	return err
}

// Chmod changes the permission bits of a virtual network.
func (vc *VirtualNetworkController) Chmod(perm shared.Permissions) error {
	return vc.ChmodContext(context.Background(), perm)
}

// ChmodContext changes the permission bits of a virtual network.
func (vc *VirtualNetworkController) ChmodContext(ctx context.Context, perm shared.Permissions) error {
	args := append([]interface{}{vc.ID}, perm.ToArgs()...)
	_, err := vc.c.Client.CallContext(ctx, "one.vn.chmod", args...)
	return err
}

// Chown changes the ownership of a virtual network.
// * userID: The User ID of the new owner. If set to -1, it will not change.
// * groupID: The Group ID of the new group. If set to -1, it will not change.
func (vc *VirtualNetworkController) Chown(userID, groupID int) error {
	return vc.ChownContext(context.Background(), userID, groupID)
}

// ChownContext changes the ownership of a virtual network.
// * ctx: context for cancelation
// * userID: The User ID of the new owner. If set to -1, it will not change.
// * groupID: The Group ID of the new group. If set to -1, it will not change.
func (vc *VirtualNetworkController) ChownContext(ctx context.Context, userID, groupID int) error {
	_, err := vc.c.Client.CallContext(ctx, "one.vn.chown", vc.ID, userID, groupID)
	return err
}

// Rename renames a virtual network.
// * newName: The new name.
func (vc *VirtualNetworkController) Rename(newName string) error {
	return vc.RenameContext(context.Background(), newName)
}

// RenameContext renames a virtual network.
// * ctx: context for cancelation
// * newName: The new name.
func (vc *VirtualNetworkController) RenameContext(ctx context.Context, newName string) error {
	_, err := vc.c.Client.CallContext(ctx, "one.vn.rename", vc.ID, newName)
	return err
}

// Lock locks the vn following lock level. See levels in locks.go.
func (vc *VirtualNetworkController) Lock(level shared.LockLevel) error {
	return vc.LockContext(context.Background(), level)
}

// LockContext locks the vn following lock level. See levels in locks.go.
func (vc *VirtualNetworkController) LockContext(ctx context.Context, level shared.LockLevel) error {
	_, err := vc.c.Client.CallContext(ctx, "one.vn.lock", vc.ID, level)
	return err
}

// Unlock unlocks the vn.
func (vc *VirtualNetworkController) Unlock() error {
	return vc.UnlockContext(context.Background())
}

// UnlockContext unlocks the vn.
func (vc *VirtualNetworkController) UnlockContext(ctx context.Context) error {
	_, err := vc.c.Client.CallContext(ctx, "one.vn.unlock", vc.ID)
	return err
}

// Recover recovers a stuck Virtual Network
// * op: (0) failure, (1) success, (2) retry, (3) delete
func (vc *VirtualNetworkController) Recover(op int) error {
	return vc.RecoverContext(context.Background(), op)
}

// RecoverContext recovers a stuck Virtual Network
// * ctx: context for cancelation
// * op: (0) failure, (1) success, (2) retry, (3) delete
func (vc *VirtualNetworkController) RecoverContext(ctx context.Context, op int) error {
	_, err := vc.c.Client.CallContext(ctx, "one.vn.recover", vc.ID, op)
	return err
}

// RecoverFailure forces a failure
func (vc *VirtualNetworkController) RecoverFailure() error {
	return vc.Recover(0)
}

// RecoverSuccess forces a success
func (vc *VirtualNetworkController) RecoverSuccess() error {
	return vc.Recover(1)
}

// RecoverDelete delete the network, call driver cleanup action
func (vc *VirtualNetworkController) RecoverDelete() error {
	return vc.Recover(2)
}
